### PR TITLE
Support CREATE/DROP FULLTEXT INDEX for fulltext search 

### PR DESCRIPTION
**Pull request by roshan0708**
### Description

Currently babelfish doesn't support fulltext index creation and drop for full text search. This commit implements the logic of create and drop fulltext index and also handles the unsupported features of the CREATE/DROP FULLTEXT INDEX statements. 

Signed-off-by: Roshan Kanwar [rskanwar@amazon.com](mailto:rskanwar@amazon.com)

### Issues Resolved

Task: BABEL-4383

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** Yes

* **Arbitrary inputs -**

* **Negative test cases -** Yes

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/guc.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c]: what is the purpose of this change?

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: should be check for schema name being NULL?

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: check for null schema to be on safe side

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: this can be combined with if else
if (get_index_name(relid)) error else get query str

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: also print index name (mostly try to match what error SQL server will give)

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: check for null schema

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: we can handle both in a single function with a boolean to distinguish create vs drop, most of the code is duplicate

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: why this assumption that get_index_name will only return full text index? what about other indexes on table? If this function is only for full text index, it should be named accordingly

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: why are we assuming that first index is full text index only?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:1947]: Please make sure column_name list and index name is not null or we handle null values

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: why no check for schema name being NULL here? also, is it guaranteed that index name is NULL across database?

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: ALTER FULLTEXT rather than ALTER CREATE FULLINDEX

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Indetnantion is messed up

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: which filegroup option we support?

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: which with options we support?

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp:6763]: getting table name could be a utility function shared between both create/drop

File: contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp]: why we have same error handling at two places?

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: generic comment
1) We should have test case for each error we are throwing, in executor or ANTLR
2) We should have test case for each condition if/else we have. For example : creating full text index on one column vs multiple column

File: contrib/babelfishpg_tsql/src/guc.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/guc.c]: This is to only allow superuser to enable full text search, initially we were using escape hatch which the customer can use to enable fts anytime which we don’t want. After pentest, if everything goes right, we will remove this.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: That is already handled in tsqlIface file, while I’m extracting other data items.

Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: Just for sanity, I think schema_name should be declared as const here, e.g. `const char *schema_name`

File: contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp]: +1, if we already are catching unsupported features here, it makes no sense to have the same error handling in the main handler. I would remove the handling in the main function and leave the error handling here.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: In the query statement, if we have not specified the schema name, it will be an empty string i.e. `""` else we get the proper schema name in the helper function. So the schema name will never be null. Here is the below code for reference - 
```
if (ctx->table_name())
{
	std::string table_info = ::getFullText(ctx->table_name());
	std::string table_name = "";
	std::string schema_name = "";
	size_t pos = table_info.find(".");
	if (pos != std::string::npos) {
		// Extract the schema name before the "."
		schema_name = table_info.substr(0, pos);
		// Extract the table name after the "."
		table_name = table_info.substr(pos + 1);
	} else {
		// No "." character found, set first to the entire string
		table_name = table_info;
	}
	stmt->table_name = pstrdup(downcase_truncate_identifier(table_name.c_str(), table_name.length(), true));
	stmt->schema_name = pstrdup(downcase_truncate_identifier(schema_name.c_str(), schema_name.length(), true));
}
```

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: We don't support any filegroup options now.

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: We don't support any with options now.

File: contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp]: Will do

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: This is specifically for fulltext index, will rename it.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: Because there can be only one fulltext index per table acc. to SQL Server docs. We get a list of indexes, so we assume we get the fulltext index in the first item only.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: Yes, I've added those test cases in the JDBC test files of full text index.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: Done

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c:1947]: We already handle null values, so if we pass no column or NULL as index name in the query statement, ANTLR will automatically throw `syntax error`.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (Sairakan) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: I'm not clear on why this function exists separately from other fully qualified schema name resolutions - can you add a comment header to the function explaining why we need this? Is full text search the only case when this kind of fully qualified name is needed?

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (Sairakan) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c:3916]: The error message mentions permissions - have we tested for cases of users that wouldn't have DROP FULLTEXT INDEX permissions

**Review by Sairakan — APPROVED** [review]
Suggestions to clean up/clarify the code, but otherwise looks good to me. I do have the question about permission checking, but that is lower priority.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: Actually, it's just a helper function that I created specifically for getting the schema name for FTS. It's basically just a combination of two functions to handle the master db schema and custom schema case.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c:3916]: Actually this is a generic sql server error message which is thrown when there exist no fulltext index to be deleted from the specified table. Although according to the official docs:

> To drop a full-text index in SQL Server, you need to have ALTER permission on the table or indexed view, or be a member of the sysadmin fixed server role, or db_owner or db_ddladmin fixed database roles.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (Sairakan) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: per https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/statements/create-fulltext-index-transact-sql.md#key-index-index_name, in addition to being unique, the index, must also be non-nullable and single-key. I am not sure if the pg_index tuple has metadata for all of these conditions, but if it does we might as well include them in this check as well.

File: contrib/babelfishpg_tsql/src/guc.c
Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/src/guc.c]: escape hatch usually means that when set corresponding commands will be ignored so that application can work. What will we do in case of FTS feature rollback? how this change will help?

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: indentation

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: we are checking for NULL schema here but not in gen_schema_name_for_fulltext_index? Either put an assert that schema name should not be NULL or add checks.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: indentation is messed up at number of places

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp:6704]: we should add test cases with mixed case schema, table and index name

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: indentation

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: if we do not support any, can we remove if and just give an error for having catalog_filegroup_option

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp
Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: same here

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c:3906]: there is no test for this, please make sure to add test case for each possible error case

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c:3916]: we should add some test cases for permission check as well, what level of permissions apply at schema, table level etc.

File: contrib/babelfishpg_tsql/src/guc.c
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/guc.c]: If a normal user tries to use this escape hatch to turn on fulltext search, it will throw `permission error`. This can only be turned on if the login is of super user. We have just put it like this so that let's say we roll FTS feature in prod, the customer shouldn't be able to use it rather only we, as superuser, should be able to turn it on. We'll rollback these extra changes of escape hatch once the pentest is done, that's what we have planned so that the customer is able to use it normally without turning on any escape hatch.

**Review by Sairakan — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/fts.c
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/fts.c:22]: Are we sure about this error msg ? It's not supported ?

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c:3923]: I'm a litte confused about this part, in some other impl, we do construct a structure, not a query string.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:1971]: It's not a good idea to use string to pass a internal real cmd.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:1891]: Do we need to error out in here ?

File: contrib/babelfishpg_tsql/src/fts.c
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/fts.c:22]: This is a fail-safe, as we are waiting for appsec review. The idea is that until the parser is approved by appsec, it will be controlled by a superuser-only GUC. Customers who wish to experiment with full text search in red panda can submit a request for CP to override the GUC to allow them to use full text search features. We will remove this GUC in the next release after Red Panda assuming that it works well.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:1891]: This is in order to make this feature work with Xiaohui's PR in https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1990 - in PG, GIN indexes don't support the NULL ordering which was implemented in that PR to apply to all index creation.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c:3923]: All other cases of calling this function that I found in the code just use a StringInfoData and call the function on the char* contained within the struct, so I don't think this is that different.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c:1971]: Agreed, in the next release I plan to rework this to just generate the correct parse tree within gram-tsql-epilogue.y.c

File: contrib/babelfishpg_tsql/src/pl_exec-2.c
Comment (forestkeeper) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c:3923]: would you mind refer to : transformAlterTableStmt , and transformSelectIntoStmt

**Review by forestkeeper — APPROVED** [review]

**Pull request merged**
